src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.Encrypt.cs

```diff
@@ -0,0 +1,364 @@
+                }
+                pEnvelopedEncodeInfo->rgCmsRecipients = rgCmsRecipients;
+
+                int numCertificates = originatorCerts.Count;
+                pEnvelopedEncodeInfo->cCertEncoded = numCertificates;
+                pEnvelopedEncodeInfo->rgCertEncoded = null;
+                if (numCertificates != 0)
+                {
+                    DATA_BLOB* pCertEncoded = (DATA_BLOB*)(hb.Alloc(numCertificates, sizeof(DATA_BLOB)));
+                    for (int i = 0; i < numCertificates; i++)
+                    {
+                        byte[] certEncoded = originatorCerts[i].Export(X509ContentType.Cert);
+                        pCertEncoded[i].cbData = (uint)(certEncoded.Length);
+                        pCertEncoded[i].pbData = hb.AllocBytes(certEncoded);
+                    }
+                    pEnvelopedEncodeInfo->rgCertEncoded = pCertEncoded;
+                }
+
+                pEnvelopedEncodeInfo->cCrlEncoded = 0;
+                pEnvelopedEncodeInfo->rgCrlEncoded = null;
+
+                pEnvelopedEncodeInfo->cAttrCertEncoded = 0;
+                pEnvelopedEncodeInfo->rgAttrCertEncoded = null;
+
+                int numUnprotectedAttributes = unprotectedAttributes.Count;
+                pEnvelopedEncodeInfo->cUnprotectedAttr = numUnprotectedAttributes;
+                pEnvelopedEncodeInfo->rgUnprotectedAttr = null;
+                if (numUnprotectedAttributes != 0)
+                {
+                    CRYPT_ATTRIBUTE* pCryptAttribute = (CRYPT_ATTRIBUTE*)(hb.Alloc(numUnprotectedAttributes, sizeof(CRYPT_ATTRIBUTE)));
+                    for (int i = 0; i < numUnprotectedAttributes; i++)
+                    {
+                        CryptographicAttributeObject attribute = unprotectedAttributes[i];
+                        pCryptAttribute[i].pszObjId = hb.AllocAsciiString(attribute.Oid.Value);
+                        AsnEncodedDataCollection values = attribute.Values;
+                        int numValues = values.Count;
+                        pCryptAttribute[i].cValue = numValues;
+                        DATA_BLOB* pValues = (DATA_BLOB*)(hb.Alloc(numValues, sizeof(DATA_BLOB)));
+                        for (int j = 0; j < numValues; j++)
+                        {
+                            byte[] rawData = values[j].RawData;
+                            pValues[j].cbData = (uint)(rawData.Length);
+                            pValues[j].pbData = hb.AllocBytes(rawData);
+                        }
+                        pCryptAttribute[i].rgValue = pValues;
+                    }
+                    pEnvelopedEncodeInfo->rgUnprotectedAttr = pCryptAttribute;
+                }
+
+                return pEnvelopedEncodeInfo;
+            }
+
+            //
+            // This returns an allocated native memory block. Its lifetime (and that of any allocated subblocks it may point to) is the shorter that of "hb". 
```

Comment:
"the shorter" doesn't make sense here.  I assume this is the normal "the memory is monitored by hb, we'll free it in a bit, I promise".  I'm not seeing how it could be shorter than hb, since it gets used in the "do everything" method.